core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java

```diff
@@ -318,6 +323,16 @@ private Object processWithExceptionHandling(Expression expression, Object contex
             }
         }
 
+        private boolean canUseCache(Object context)
+        {
+            // We are using WeakHashMap as cache, that's why we can't depend on an intermediate object
+            // that consists of expression as well as context as a key. This is because intermediate object could
+            // get GCed even when expression still exists which will cause cache miss. So, to solve this problem we
+            // can only cache when context is noop. It'll also give us pretty good perf benefits because in most of
+            // places optimize is called with noop symbol resolver.
+            return optimize && context instanceof NoOpSymbolResolver;
```

Comment:
it shouldn't matter if we `optimize` or not, should it?

Reply:
I think it matters. In some places, we are returning different results in case of optimize. For instance, in case of function calls, we don't optimize if the function isn't deterministic. 

https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/sql/planner/ExpressionInterpreter.java#L1057

Reply:
Agreed, the class has two 'modes' and we cannot uniformly cache for both of them at the same time, unless 'the mode' becomes a cache key too